hikari/internal/data_binding.py

```diff
@@ -75,62 +76,131 @@
 Stringish = typing.Union[str, int, bool, undefined.UndefinedType, None, snowflakes.Unique]
 """Type hint for any valid that can be put in a StringMapBuilder"""
 
+JSONEncoder = typing.Callable[[typing.Union[JSONArray, JSONObject]], typing.Union[str, bytes]]
+"""Type hint for hikari-compatible JSON encoders.
+
+A hikari-compatible JSON encoder is one which will take in a JSON-ish object and output either `str`
+or `bytes`.
+"""
+
+JSONDecoder = typing.Callable[[typing.Union[str, bytes]], typing.Union[JSONArray, JSONObject]]
+"""Type hint for hikari-compatible JSON decoder.
+
+A hikari-compatible JSON decoder is one which will take either a `str` or `bytes` and outputs
+the JSON-ish object, as well as raises a `ValueError` on an incorrect JSON payload being passed in.
+"""
+
 _StringMapBuilderArg = typing.Union[
     typing.Mapping[str, str], multidict.MultiMapping[str], typing.Iterable[typing.Tuple[str, str]]
 ]
 
 _APPLICATION_OCTET_STREAM: typing.Final[str] = "application/octet-stream"
+_JSON_CONTENT_TYPE: typing.Final[str] = "application/json"
+_UTF_8: typing.Final[str] = "utf-8"
 
-if typing.TYPE_CHECKING:
-    JSONDecodeError: typing.Type[Exception] = Exception
-    """Exception raised when loading an invalid JSON string."""
+default_json_dumps: JSONEncoder
+"""Default json encoder to use."""
 
-    def dump_json(_: typing.Union[JSONArray, JSONObject], /, *, indent: int = ...) -> str:
-        """Convert a Python type to a JSON string."""
-        raise NotImplementedError
+default_json_loads: JSONDecoder
+"""Default json decoder to use."""
 
-    def load_json(_: typing.AnyStr, /) -> typing.Union[JSONArray, JSONObject]:
-        """Convert a JSON string to a Python type."""
-        raise NotImplementedError
+try:
+    import orjson
 
-else:
+    default_json_dumps = orjson.dumps
+    default_json_loads = orjson.loads
+except ModuleNotFoundError:
     import json
 
-    dump_json = json.dumps
-    """Convert a Python type to a JSON string."""
+    def default_json_dumps(obj: typing.Union[JSONArray, JSONObject]) -> str:
+        """Encode a JSON object to a `str`."""
+        return json.dumps(obj, separators=(",", ":"))
+
+    default_json_loads = json.loads
+
+
+@typing.overload
+def dump_json(  # noqa: D103 - Missing docstring
+    obj: typing.Union[JSONArray, JSONObject],
+    /,
+    *,
+    encode: typing.Literal[False],
+    json_dumps: JSONEncoder = default_json_dumps,
+) -> str:
+    ...
+
+
+@typing.overload
+def dump_json(  # noqa: D103 - Missing docstring
+    obj: typing.Union[JSONArray, JSONObject],
+    /,
+    *,
+    encode: typing.Literal[True],
+    json_dumps: JSONEncoder = default_json_dumps,
+) -> bytes:
+    ...
 
-    load_json = json.loads
-    """Convert a JSON string to a Python type."""
 
-    JSONDecodeError = json.JSONDecodeError
-    """Exception raised when loading an invalid JSON string."""
+def dump_json(
```

Comment:
this whole function just feels like a unnecessary extra steps ngl (esp considering the whole point of this is to plop down into pretty important hot paths). Are there even any cases where encode isn't True cause if not then this is just unnecessary complexity?

Reply:
I dont like it either, but in the hopes of providing a more agnostic interface, we need to ensure that the deserializers that we get passed behave as we expect them to (thank you orjson for being the special one!)

Reply:
It's not that hard for ppl to do `dumps=lambda d: dumps(d).encode()`

Reply:
Sorted!